### ISSUE TITLE
Outlook on the web: custom property not saved in onsend script

### ISSUE THREAD

**Issue by amiram** [OfficeDev/office-js#1019]
when you set properties in onsend script, the last property you set is not being saved. I have an onsend script that saves some custom data in the mail item metadata. After each property change I save the custom property and wait until the save callback called with success. All data is saved except from the last one. If I repeat the last property set twice, it works. The bug doesn't exist in windows or mac, just in the web version.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 6df38293-7e0a-a019-ded2-3ed960ceb18b
* Version Independent ID: 62a8ffba-8baa-c6ff-4063-81b507fc8dd4
* Content: [Get and set metadata in an Outlook add-in - Office Add-ins](https://docs.microsoft.com/en-us/office/dev/add-ins/outlook/metadata-for-an-outlook-add-in?view=outlook-js-preview#feedback)
* Content Source: [docs/outlook/metadata-for-an-outlook-add-in.md](https://github.com/OfficeDev/office-js-docs-pr/blob/master/docs/outlook/metadata-for-an-outlook-add-in.md)
* Product: **outlook**
* Technology: **add-ins**
* GitHub Login: @o365devx
* Microsoft Alias: **o365devx**

**Comment by exextoc** [comment]
Can you share the code snippet which you are trying?

**Comment by amiram** [comment]
Some code:
We have a wrapper around mail item.
```
class MailItem {
    constructor(private item: MessageCompose | MessageRead,
                protected customProperties: CustomProperties) {
    }

    protected promisify<T>(fn: (callback: (asyncResult: Office.AsyncResult<T>) => void) => void, context?): () => Promise<T>;
    protected promisify<T, P>(fn: (p: P, callback: (asyncResult: Office.AsyncResult<T>) => void) => void, context?): (p: P) => Promise<T>;
    protected promisify<T, P1, P2>(fn: (p1: P1, p2: P2, callback: (asyncResult: Office.AsyncResult<T>) => void) => void, context?): (p1: P1, p2: P2) => Promise<T>;
    protected promisify(fn, context?) {
        return (...params) => {
            return new Promise((resolve, reject) => {
                const cb = (asyncResult: AsyncResult<any>) => {
                    if (asyncResult.status === Office.AsyncResultStatus.Succeeded) {
                        resolve(asyncResult.value);
                    } else {
                        reject(asyncResult.error);
                    }
                };
                try {
                    return fn.bind(context || this.item)(...[...params, cb]);
                } catch (err) {
                    reject(err);
                }
            });
        };
    }

    protected getCustomProperty(name: string) {
        return this.customProperties.get(name);
    }

    protected setCustomProperty(name: string, value) {
        if (!value && value !== 0) {
            this.customProperties.remove(name);
        } else {
            this.customProperties.set(name, value.toString());
        }

        return this.promisify(this.customProperties.saveAsync, this.customProperties)();
    }

    get someProp(): number {
        return this.getCustomProperty('someProp');
    }

    setSomeProp(value: number) {
        return this.setCustomProperty('someProp', value);
    }
}
```
Now this wrapper is created after waiting for the custom properties to load:
```
function _getMailItem() {
    return new Promise<{ item: MessageCompose & MessageRead, customProperties: CustomProperties }>(async (resolve, reject) => {
        await Office.onReady();

        const item = Office.context.mailbox.item;
        if (item) {
            item.loadCustomPropertiesAsync(result => {
                if (result.status === Office.AsyncResultStatus.Succeeded) {
                    resolve({item: Office.context.mailbox.item, customProperties: result.value});
                } else {
                    reject(new Error('Load custom properties failed'));
                }
            });
        } else {
            resolve();
        }
    });
}

export async function getMailItem() {
    const item = await _getMailItem();
    return item && new MailItem(item.item, item.customProperties);
}
```
We have an add-in command for onsend event:
```
async function onSend(event: Event) {
    await Office.onReady();
    const mailItem = await getMailItem();
    // Now all custom properties are loaded
    await mailitem.setSomeProp(3);
    event.completed({allowEvent: true});
}
```
After the message is sent we try to read the custom properties and see if we should show some data about the message:
```
if (mailitem.someProp) {
   // Show some data in the sidebar
}
```
This is working fine in outlook on Windows and Mac, but in the web version, some other properties  are set but someProp which is set last is empty, although when you try to read it right after the set in the same method you get the value. I also verified that with OutlookSpy. The workaround was to set it twice - just duplicating the row.

**Comment by exextoc** [comment]
We aren't able to reproduce the bug. We need some more information regarding the bug. 

- After the mail is sent, are you checking the custom properties in senders sent box or receivers inbox ? Can you please check for both the scenarios. 

- Can you check the logs while the mail sent is happening for any error (if any) ?

- Is the bug consistent for you ? Does it happen every time you send a mail with onSend and custom property set ?

**Comment by clavx** [comment]
The same thing happened to me.

After I created and sent an email with an attachment, there is no custom property.
However, it doesn't always reproduce, so we can't identify the cause.

The procedure for reproduction is as follows:
1. Open the Outlook on the web
2. Create a new email
3. Attaching an image file of about 30 KB or a PDF file of about 4 MB
4. Send
5. Repeat step 2-4 several times.


Here is the reproduced code:
``` js
let _customProps;
const setCustomProp = (name, value) => {
  return new Promise((resolve) => {
    _customProps.set(name, value);
    _customProps.saveAsync(() => {
      resolve();
    });
  });
}

const prepend = (data) => {
  return new Promise((resolve) => {
    Office.context.mailbox.item.body.prependAsync(data, { coercionType: "html" }, (result) => {
      resolve(result.value);
    });
  });
}

const getAttachments = () => {
  return new Promise((resolve) => {
    Office.context.mailbox.item.getAttachmentsAsync((result) => {
      resolve(result.value);
    });
  });
}

const removeAttachment = (attachmentId) => {
  return new Promise((resolve) => {
    Office.context.mailbox.item.removeAttachmentAsync(attachmentId, (result) => {
      resolve(result.value);
    });
  });
}
```
``` js
const mainAction = (event: Office.AddinCommands.Event) => {
  Office.context.mailbox.item.loadCustomPropertiesAsync(async (result) => {
    _customProps = result.value;
    const attachments = await getAttachments();
    const a = () => removeAttachment(attachments[0].id);
    const b = () => prepend("prepend");
    const c = () => setCustomProp("setCustomProp", "setCustomProp");
    await Promise.all([ a(), b(), c() ]);
    event.completed({allowEvent: true});
  });
};
```

**Comment by amiram** [comment]
@clavx thanks for your comment. This bug is not consistent for me also. I don't see any error when it happened. Try to set it twice like:
```
await Promise.all([ a(), b(), c() ]);
await c();
```
See if it helps.

**Comment by clavx** [comment]
@amiram Thank you for your advice.

By setting the custom property twice, we confirmed that the custom property was set.

However, I noticed that sometimes the attachment is not deleted.

![attachments_remain](https://user-images.githubusercontent.com/58804032/77891516-0ad69d80-72ac-11ea-842d-d8cf467be641.gif)

When I executed the reproduction code, the custom property was set when the attachment was not deleted. It does not always happen, so the cause is still unknown.

**Comment by amiram** [comment]
@clavx  Did you try to do the async operation one by one?
await a();
await b();
await c();

**Comment by clavx** [comment]
@amiram I tried.
In that case, there was no problem.
The attachment has been deleted and the custom property have been set.

Suspecting some problems with `Promise.all` processing, I executed the following code.
However, a similar phenomenon occurred.

``` js
const mainAction = (event: Office.AddinCommands.Event) => {
  let counter = 0;
  const cb = () => {
    if (++counter === 3) {
      event.completed({allowEvent: true});
    }
  };

  Office.context.mailbox.item.loadCustomPropertiesAsync((result1) => {
    Office.context.mailbox.item.getAttachmentsAsync((result2) => {
      const attachments = result2.value;
      Office.context.mailbox.item.removeAttachmentAsync(attachments[0].id, cb);
      Office.context.mailbox.item.body.prependAsync("prepend", { coercionType: "html" }, cb);
      const _customProps = result1.value;
      _customProps.set("setCustomProp", "setCustomProp");
      _customProps.saveAsync(cb);
    });
  });
};
```

Is there any problem with the timing of the event loop in the browser?

**Comment by amiram** [comment]
@clavx I don't think it is a javascript problem. I assume performing multiple operations on the mail item may cause some problems. If each operation is getting the item state, change it and then save, all operations get the same state but each one is saving only each own change. I advise you to avoid this and wait for each operation to complete before you start another.

**Comment by clavx** [comment]
@amiram I see. You are right.

There was a difference in the behavior between the Windows version and the Web version after calling `event.completed`. In the Windows version, after calling `event.completed`, it does not terminate immediately, and it seems to have a little extra time. In the Web version, after calling `event.completed`, it terminated immediately.
So I extended the end of the program by 5 seconds: `setTimeout(() => event.completed({allowEvent: true}), 5000);` As a result, I was able to see that something was blocking when there was a problem with the attachment not being deleted. The reason why the attachment remains is that it is supposed to be deleted, but it remains on the screen.

![attachments_remain_block](https://user-images.githubusercontent.com/58804032/78107347-469a7000-7430-11ea-96dc-7d35a9c002d9.gif)

After all, there seems to be some kind of bug when performing multiple operations on asynchronous functions. I think something is happening even if the custom property is not set.

For the time being,  I'll try to make the asynchronous functions to run synchronously one at a time.

Thank you.
